### PR TITLE
Retrieving Secondary User Stores VIA Admin Services.

### PR DESCRIPTION
**Pull request by Kamidu**
Enable the Jaggry app to communicate with the admin services in order to retrieve secondary user stores available.

### CODE REVIEW

File: modules/apps/jaggery/mdm/src/mdm/config/config.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
   "httpURL": "%http.ip%",
   "iOSConfigRoot": "%https.ip%/ios-enrollment/",
   "iOSAPIRoot": "%https.ip%/ios/",
+  "carbonServer": "https://10.100.7.102:9443/",
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/config/config.json]: Some IP seems to have been hard-coded. Please fix.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 2]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+    /**
+     * Private method to be used by addUser() to
+     * retrieve secondary user stores.
+     * This needs Authentication since the method access admin services.
+     *
+     * @returns {string array} Array of secondary user stores.
+     */
+    publicMethods.getSecondaryUserStores = function () {
+        var ws = require('ws');
+    	var headers = [{name : "Authorization", value : "Basic YWRtaW46YWRtaW4="}];
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Code lines are not properly aligned it seems. + Can see some hard-coded Authorisation header value in this particular line.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 3]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+        var wsPayload = "<xsd:getSecondaryRealmConfigurations  xmlns:xsd='http://org.apache.axis2/xsd'/>";
+        var wsResponse;
+        try {
+    	    wsRequest.open(options, mdmProps.carbonServer + "services/UserStoreConfigAdminService.UserStoreCon" +
+    	        "figAdminServiceHttpsSoap12Endpoint/", false);
+    	    wsRequest.send(wsPayload);
+    	    wsResponse = wsRequest.responseE4X;
+        } catch (e) {
+    		log.error("Carbon server error while retrieving secondary user stores :" + e.toString());
+    	}
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Code is not formatted properly. Please fix.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 4]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+        options.action = "urn:getSecondaryRealmConfigurations";
+        var wsPayload = "<xsd:getSecondaryRealmConfigurations  xmlns:xsd='http://org.apache.axis2/xsd'/>";
+        var wsResponse;
+        try {
+    	    wsRequest.open(options, mdmProps.carbonServer + "services/UserStoreConfigAdminService.UserStoreCon" +
+    	        "figAdminServiceHttpsSoap12Endpoint/", false);
+    	    wsRequest.send(wsPayload);
+    	    wsResponse = wsRequest.responseE4X;
+        } catch (e) {
+    		log.error("Carbon server error while retrieving secondary user stores :" + e.toString());
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Exception trace is swallowed it seems. Please fix.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 5]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+    	}
+    	var domainIDs = stringify(wsResponse.*::['return'].*::domainId.text());
+    	if (domainIDs != "\"\"") {
+    	    var characterNeedReplacing = "\"";
+            var regExpForSearch = new RegExp(characterNeedReplacing, "g");
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Do we need to instantiate a regExp per invocation? Or can we cache one and re-use?

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 6]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+    	    wsRequest.open(options, mdmProps.carbonServer + "services/UserStoreConfigAdminService.UserStoreCon" +
+    	        "figAdminServiceHttpsSoap12Endpoint/", false);
+    	    wsRequest.send(wsPayload);
+    	    wsResponse = wsRequest.responseE4X;
+        } catch (e) {
+    		log.error("Carbon server error while retrieving secondary user stores :" + e.toString());
+    	}
+    	var domainIDs = stringify(wsResponse.*::['return'].*::domainId.text());
+    	if (domainIDs != "\"\"") {
+    	    var characterNeedReplacing = "\"";
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: If this is a constant, then you should perhaps follow proper constant declaration standards.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 7]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+     *
+     * @returns {string array} Array of secondary user stores.
+     */
+    publicMethods.getSecondaryUserStores = function () {
+        var ws = require('ws');
+    	var headers = [{name : "Authorization", value : "Basic YWRtaW46YWRtaW4="}];
+        var wsRequest = new ws.WSRequest();
+        var options = [];
+        options.HTTPHeaders = headers;
+        options.useSOAP = 1.2;
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Use appropriate constants for magic numbers like these.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js [hunk 8]
```diff
@@ -540,5 +540,42 @@ var userModule = function () {
...
+        options.useWSA = 1.0;
```
Comment (prabathabey) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Same goes here as well.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 2 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Code Line are formatted.formatted code can be seen when viewing the full file.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 3 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: Code Line are formatted.formatted code can be seen when viewing the full file.

File: modules/apps/jaggery/mdm/src/mdm/modules/user.js (hunk 5 above)
Comment (Kamidu) [inline modules/apps/jaggery/mdm/src/mdm/modules/user.js]: the regExp only instantiate on the page load.and this is regExp is not a huge resource hungry complicated one.As for the discussion had offline this kept unchanged.

**Pull request closed without merging**